google-cloud-pubsub/src/test/java/com/google/cloud/pubsub/v1/PublisherImplTest.java

```diff
@@ -923,6 +928,191 @@ public void testShutDown() throws Exception {
+  }
+
+  @Test
+  public void testPublishFlowControl_throwExceptionWithOrderingKey() throws Exception {
+    Publisher publisher =
+        getTestPublisherBuilder()
+            .setExecutorProvider(SINGLE_THREAD_EXECUTOR)
+            .setBatchingSettings(
+                Publisher.Builder.DEFAULT_BATCHING_SETTINGS
+                    .toBuilder()
+                    .setElementCountThreshold(1L)
+                    .setDelayThreshold(Duration.ofSeconds(5))
+                    .setFlowControlSettings(
+                        FlowControlSettings.newBuilder()
+                            .setLimitExceededBehavior(
+                                FlowController.LimitExceededBehavior.ThrowException)
+                            .setMaxOutstandingElementCount(1L)
+                            .setMaxOutstandingRequestBytes(10L)
+                            .build())
+                    .build())
+            .setEnableMessageOrdering(true)
+            .build();
+
+    // Sending a message that is too large results in an exception.
+    ApiFuture<String> publishFuture1 =
+        sendTestMessageWithOrderingKey(publisher, "AAAAAAAAAAA", "a");
+    try {
+      publishFuture1.get();
+      fail("Should have thrown an FlowController.MaxOutstandingRequestBytesReachedException");
+    } catch (ExecutionException e) {
+      assertThat(e.getCause())
+          .isInstanceOf(FlowController.MaxOutstandingRequestBytesReachedException.class);
+    }
+
+    // Sending a second message for the same ordering key fails because the first one failed.
+    ApiFuture<String> publishFuture2 = sendTestMessageWithOrderingKey(publisher, "AAAA", "a");
+    try {
+      publishFuture2.get();
+      Assert.fail("This should fail.");
+    } catch (ExecutionException e) {
+      assertEquals(SequentialExecutorService.CallbackExecutor.CANCELLATION_EXCEPTION, e.getCause());
+    }
+  }
+
+  @Test
+  public void testPublishFlowControl_block() throws Exception {
+    final Publisher publisher =
+        getTestPublisherBuilder()
+            .setExecutorProvider(SINGLE_THREAD_EXECUTOR)
+            .setBatchingSettings(
+                Publisher.Builder.DEFAULT_BATCHING_SETTINGS
+                    .toBuilder()
+                    .setElementCountThreshold(1L)
+                    .setDelayThreshold(Duration.ofSeconds(5))
+                    .setFlowControlSettings(
+                        FlowControlSettings.newBuilder()
+                            .setLimitExceededBehavior(FlowController.LimitExceededBehavior.Block)
+                            .setMaxOutstandingElementCount(2L)
+                            .setMaxOutstandingRequestBytes(10L)
+                            .build())
+                    .build())
+            .build();
+    Executor responseExecutor = Executors.newScheduledThreadPool(10);
+    final CountDownLatch sendResponse1 = new CountDownLatch(1);
+    final CountDownLatch sentResponse1 = new CountDownLatch(1);
```

Comment:
nit. can we rename sentResponse1 to response1Sent or something similar.  sendResponse1 and sentResponse1 are very close in name.